### PR TITLE
[Toolchain] Add toolchain tree provider

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit adds toolchain tree provider.

ONE-vscode-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -64,9 +65,10 @@ export function activate(context: vscode.ExtensionContext) {
   new OneExplorer(context, logger);
 
   // ONE view
-  let refreshCompiler = vscode.commands.registerCommand('onevscode.refresh-toolchain', () => {
-    console.log('refresh-toolchain: NYI');
-  });
+  const toolchainProvier = new ToolchainProvider();
+  vscode.window.registerTreeDataProvider('ToolchainView', toolchainProvier);
```
Comment (dayo09) [inline src/extension.ts:69]: The registered treeDataProvider must be subscribed to be disposed after the extension's expiration.
```suggestion
  context.subscriptions.push(
    vscode.window.registerTreeDataProvider('ToolchainView', toolchainProvier));
```

FYI, https://github.com/Samsung/ONE-vscode/issues/643

**Comment by yunjayh** [comment]
Format error T^T

File: src/Toolchain/ToolchainProvider.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      public readonly label: string,
+      public readonly collapsibleState: vscode.TreeItemCollapsibleState,
+      public readonly type: NodeType,
+      public readonly version?: string,
+  ) {
+    super(label, collapsibleState);
+
+    if (type === NodeType.backend) {
+      this.iconPath = new vscode.ThemeIcon('bracket');
+    } else if (type === NodeType.toolchain) {
```
Comment (hyunsik-yoon) [inline src/Toolchain/ToolchainProvider.ts:36]: (optional) "ToolchainNode type can be of toolchain or _backend_" could be confusing. 
How about renaming `ToolchainNode` to `ToolchainViewNode` ?

File: src/Toolchain/ToolchainProvider.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    } else {
+      return new ToolchainNode(name, vscode.TreeItemCollapsibleState.Expanded, type);
+    }
+  }
+
+  private getNode(node: ToolchainNode): ToolchainNode[] {
+    if (node.type === NodeType.backend) {
+      if (node.label in gToolchainEnvMap) {
+        const toolchain = gToolchainEnvMap[node.label].listInstalled();
+        if (toolchain && toolchain.info.version) {
```
Comment (hyunsik-yoon) [inline src/Toolchain/ToolchainProvider.ts:77]: Can I ask a question?
Current code of `listInstalled()` seems to return `undefined` even though toolchain was installed by user.

https://github.com/Samsung/ONE-vscode/blob/911636d21dadca657edd8175913c5b7424d6bdcc/src/Toolchain/ToolchainEnv.ts#L116-L118


In such case shouldn't we check if there's any toolchain already installed?

File: src/Toolchain/ToolchainProvider.ts (hunk 2 above)
Comment (jyoungyun) [reply src/Toolchain/ToolchainProvider.ts:36]: This variable always uses with `NodeType`. So I think it seems unnecessary that the node is attached to the name again.

File: src/Toolchain/ToolchainProvider.ts (hunk 3 above)
Comment (jyoungyun) [reply src/Toolchain/ToolchainProvider.ts:77]: I already checked the undefine variable.

> In such case shouldn't we check if there's any toolchain already installed?

I don't understand what you mean..

File: src/Toolchain/ToolchainProvider.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      public readonly type: NodeType,
+      public readonly version?: string,
+  ) {
+    super(label, collapsibleState);
+
+    if (type === NodeType.backend) {
+      this.iconPath = new vscode.ThemeIcon('bracket');
+    } else if (type === NodeType.toolchain) {
+      this.iconPath = new vscode.ThemeIcon('circle-filled');
+      this.label = `${this.label} ${this.version}`;
```
Comment (dayo09) [inline src/Toolchain/ToolchainProvider.ts:38]: This line may removed by 
```js
super(`${label} ${version}`, collapsibleState);
```

File: src/Toolchain/ToolchainProvider.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    if (element) {
+      return Promise.resolve(this.getNode(element));
+    } else {
+      const nodes = Object.keys(gToolchainEnvMap)
+                        .map((backend) => this.toToolchainNode(NodeType.backend, backend));
+      return Promise.resolve(nodes);
+    }
+  }
+
+  private toToolchainNode(type: NodeType, name: string, version?: string): ToolchainNode {
```
Comment (dayo09) [inline src/Toolchain/ToolchainProvider.ts:65]: How about merging this into ToolchainNode constructor?

File: src/Toolchain/ToolchainProvider.ts (hunk 4 above)
Comment (jyoungyun) [reply src/Toolchain/ToolchainProvider.ts:38]: No, it can't.
The label value is different by type.

File: src/Toolchain/ToolchainProvider.ts (hunk 5 above)
Comment (jyoungyun) [reply src/Toolchain/ToolchainProvider.ts:65]: I will do it later.

File: src/Toolchain/ToolchainProvider.ts (hunk 4 above)
Comment (dayo09) [reply src/Toolchain/ToolchainProvider.ts:38]: Sure,.

File: src/Toolchain/ToolchainProvider.ts (hunk 5 above)
Comment (dayo09) [reply src/Toolchain/ToolchainProvider.ts:65]: `toToolchainNode` sounds like a mapper. 
It's optional anyways.

**Review by dayo09 — APPROVED** [review]

File: src/Toolchain/ToolchainProvider.ts (hunk 5 above)
Comment (jyoungyun) [reply src/Toolchain/ToolchainProvider.ts:65]: Yes, you're right. This function will be put into another function with other functions implemented.

File: src/Toolchain/ToolchainProvider.ts (hunk 3 above)
Comment (hyunsik-yoon) [reply src/Toolchain/ToolchainProvider.ts:77]: Ah. I realized that `ToolchainEnv.installed` was _already_ set when ONE-vscode starts to run. Sorry for confusion.

**Review by hyunsik-yoon — APPROVED** [review]

**Pull request merged**